### PR TITLE
Add env variable matrix to pass to tests and build

### PR DESCRIPTION
**Pull request by Alphare**
Hello,

This PR introduces support for environment variables, as askied in #165 .
When discussing the feature during the [Mercurial mini-sprint](https://octobus.net/blog/2019-04-17-mercurial-mini-sprint-2019-04.html) we held in Paris earlier this month, it became apparent that there would be a need for specifying whether or not a given environment variable should result in an additional build. Thus, environment variables can be defined as being "build" or "non_build". 
This decision has shaken up at least of core feature of ASV: how to identify an `Environment`, in that multiple `Environment`s can henceforth refer to the same build on the file system, allowing for quicker build time and less disk usage waste.
 
This PR is very close to completion (5 tests failing... I haven't been able to figure out why as of yet), but since I'm most likely not going to have time to finish it for a few weeks I figured I'd create it to kick-start the discussion that I'm sure will need to happen.

Closes: #165

### CODE REVIEW

**Comment by pv** [comment]
I'd perhaps prefer to make the split below Environment instead of above it: i.e. separate EnvironmentConfiguration (runtime configuration including specific environment variables) from Environment (the conda/virtualenv directory), instead of special-casing the handling of environment variables. Different EnvironmentConfiguration instances could use the same Environment.

This could also enable marking packages in the `matrix` as "non-build". Quite often with Python packages, the wheel can be built also without the required packages, and the build result is independent of them. If the build step is slow (e.g. c++ code), this can be useful.

**Comment by pv** [comment]
Ok, I read it now more carefully, and the overall implementation approach seems fine. The association of Environment with an unique installation directory is less important than their association with specific runtime environment, and we can well relax the former. Since environment setup is idempotent, there's no real need to split the class.

One place that needs more careful handling is the parallel creation of environments. For simplicity, we could scrap the parallelism here, as I doubt it's very useful in practice since environments are long-lived.

Minor comments:
- `env_vars_combination` -> `env_vars` (the `requirements` option is not called `requirements_combination` either)
- `_test_` -> `_benchmark_` or `_bench_` (benchmarks are not tests)
- Please avoid whitespace changes in parts of code you are not touching.

Style comments:
- Linewidths up to 120 are OK.
- Splitting things such as `get_*_env_vars` to tiny functions makes it more difficult to read --- one function with kwargs seems preferable here.
- For tests with data sets, not using `pytest.mark.parametrize` is preferable --- for loop is better.

**Comment by pv** [comment]
Rebase + minor refactoring + small bugfixes

**Comment by Alphare** [comment]
Thanks for your detailed thoughts and your help.

Indeed, marking some dependencies as non-build could be very useful as well. Not sure if we want to introduce that in this PR or in a separate one, though.

A few comments:

- Why do you say that a for loop is preferable to `parametrize`? I only see a (subjective) syntactic readability issue with `parametrize`, otherwise I think it presents clear advantages, a few examples: separate nodes in `pytest`, test output is easier to read, tests don't stop at the first failure in the set.
- I think building environments in parallel is still very important. At Octobus, we lose a significant amount of time building environments, the faster the better.
- I'm not sure that the current file naming scheme is viable. In fact, this change makes it significantly worse by appending environment variables. Windows' 256 path length limitation will be hit very easily, and it kind of gives the impression of a half-baked solution for storing metadata.

If you think I should tackle a specific task let me know. Otherwise I don't want to start working on the same thing as you.

File: asv/environment.py [hunk 1]
```diff
@@ -327,14 +327,11 @@ def get_environments(conf, env_specifiers, verbose=True):
...
 
-        env_matrix_combinations = iter_env_matrix_combinations(conf.env_matrix)
-
-        if not env_matrix_combinations:
-            env_matrix_combinations = [{}]
-
         for requirements in requirements_iter:
             python = requirements.pop('python')
 
+            env_matrix_combinations = iter_env_matrix_combinations(conf.env_matrix)
```
Comment (Alphare) [inline asv/environment.py]: Putting this iterable outside of the loop was deliberate since it is invariant. Is there something I'm missing?

**Comment by pv** [comment]
Separate PR is probably better for the rest of the matrix.

I think the biggest issue at this point would be to actually test this in real use and see if problems appear.

For parametrize, readability counts: fitting the function in one screen vertically, and the dataset goes into wrong place with the decorator (above test function name). The test cases here also seemed too small for separate nodes to me, and stopping on first failure is probably not a problem here, as the code is deterministic. Matters of taste of course.

The parallelization was easy to fix, so that's now taken care of.

The file name switches to using the env hash if the name becomes too long.

**Comment by pv** [comment]
It's a generator, so you don't get a new iterator for it each time the for loop starts, unlike for a list.

File: asv/environment.py (hunk 1 above)
Comment (Alphare) [reply asv/environment.py]: Ha, of course "it's a generator".

**Comment by pv** [comment]
I'm probably not going to touch this before the weekend, so if you want to make changes, please feel free.

**Pull request merged**